Add option to skip dropping temporary tables upon checksum failure

shixuan-fan opened the pull request:
Address #8301

electrum commented on presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java:
```diff
@@ -345,29 +341,59 @@ private QueryResult executeQueryControl()
+            if (tearDownResult.getState() == State.SUCCESS) {
+                break;
+            }
+            try {
```
Add a log message here, something like `"Query failed on attempt #1, will sleep and retry"`

electrum commented on presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java:
```diff
@@ -345,29 +341,59 @@ private QueryResult executeQueryControl()
+        int retry = 0;
+        QueryResult tearDownResult;
+        do {
+            tearDownResult = tearDown(queryPair.getTest(), testPostQueryResults, testPostquery -> executeQuery(testGateway, testUsername, testPassword, queryPair.getTest(), testPostquery, testTimeout, sessionProperties));
```
This is too long to easily read, so please wrap the arguments

electrum commented on presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java:
```diff
@@ -345,29 +341,59 @@ private QueryResult executeQueryControl()
+        return tearDownResult;
+    }
+
+    private QueryResult tearDownQueryControl()
```
These methods are basically same. Add a helper method that they both call. The method will be easier to read since the variable names will be shorter (no test/control in all the names) and it will be easier to verify that we used the correct test/control variables for each one.

electrum commented on presto-verifier/src/main/java/com/facebook/presto/verifier/VerifierConfig.java:
```diff
@@ -83,6 +83,7 @@
     private boolean shadowWrites = true;
     private String shadowTestTablePrefix = "tmp_verifier_";
     private String shadowControlTablePrefix = "tmp_verifier_";
+    private boolean skipDroppingTemporaryTables = true;
```
It's better to have the option be "positive" rather than "negative". With a negative option, setting it to false creates a double negative which is harder to reason about. Thus, this option should be `dropTemporaryTables` so that the result of true/false is more clear.

electrum commented on presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java:
```diff
@@ -227,7 +230,14 @@ private boolean validate()
-            return checkForDeterministicAndRerunTestQueriesIfNeeded();
+            matches = checkForDeterministicAndRerunTestQueriesIfNeeded();
+        }
+        if (matches && skipDroppingTemporaryTables) {
```
This logic seems to be inverted

shixuan-fan commented:
@electrum Comment addressed. Please review when you have bandwidth :)

electrum commented on presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java:
```diff
@@ -227,7 +230,28 @@ private boolean validate()
-            return checkForDeterministicAndRerunTestQueriesIfNeeded();
+            matches = checkForDeterministicAndRerunTestQueriesIfNeeded();
+        }
+        if (matches && !dropTemporaryTables) {
```
I think this should be
```java
if (matches && dropTemporaryTables)

shixuan-fan replied:
I think the logic here is that: If checksum succeeds and we didn't drop temporary tables before, we should drop them then.

electrum commented on presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java:
```diff
@@ -227,7 +230,28 @@ private boolean validate()
+                    testTimeout,
+                    testPostQueryResults,
+                    testTeardownRetries);
+            controlResult = controlTearDownResult.getState() == State.SUCCESS ? controlResult : controlTearDownResult;
```
It might be better to just print a message if teardown fails rather than failing the result of validation.

electrum commented on presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java:
```diff
@@ -341,29 +365,44 @@ private QueryResult executeQueries(
+                break;
+            }
+            try {
+                TimeUnit.MINUTES.sleep(1);
```
Print a message here so we know it is sleeping rather than just being hung. Include the attempt number

electrum commented on presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java:
```diff
@@ -341,29 +365,44 @@ private QueryResult executeQueries(
+            List<QueryResult> postQueryResults,
+            int teardownRetries)
+    {
+        int retry = 0;
```
Let's rename this to `attempt` since it's actually the attempt number rather than the number of retries

electrum commented on presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java:
```diff
@@ -341,29 +365,44 @@ private QueryResult executeQueries(
+            }
+            retry++;
+        }
+        while (retry < teardownRetries);
```
This loop structure is bad, since we sleep even if this is the last iteration (we should only sleep between attempts)

electrum commented on presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java:
```diff
@@ -341,29 +365,44 @@ private QueryResult executeQueries(
-                    break;
-                }
-                retry++;
+            if (dropTemporaryTables) {
```
The config option is for dropping on a checksum failure, but we still want to drop tables if the query fails.

shixuan-fan replied:
This is a good point. Now the validator will not drop tables when executing, and will only drop them when `validate()` returns if necessary (the only case that these tables will be kept is where checksum fails and the config is set to false).

wenleix commented on presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java:
```diff
@@ -296,59 +296,55 @@ private boolean checkForDeterministicAndRerunTestQueriesIfNeeded()
+                controlTeardownRetries);
+    }
+
+    private QueryResult executeQueries(
```
nit: `executeQuery`

shixuan-fan replied:
`executeQuery` doesn't seem right because we already have one. This is more like executing a query with setup and teardown, but I can't really come up with a name for it.

shixuan-fan commented:
@wenleix: Only when we added skip dropping will we need to extract `tearDown` out as a function, so that part doesn't really look like a refactoring in my opinion. As for logs, if I added in the first commit, it will be removed and then added again as we extracting `tearDown` out, which might not be good for review I guess?

mbasmanova commented:
@electrum David, could you take another look at this change. It would help a lot with manual re-verification of queries that didn't pass the verifier.

shixuan-fan commented:
Rebased. Feel free to take a look @electrum, @haozhun, @wenleix, @mbasmanova 

@zaynoon If we don't specify retention, it should create a table with default retention, right? I don't know if verifier is only used for Hive so I'm not sure if retention makes sense for other connector.

haozhun commented on presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java:
```diff
@@ -208,32 +214,74 @@ private boolean validate()
         // query has too many rows. Consider blacklisting.
```
rename `executeQueryControl` to `executePreAndMainForControl`

Accordingly, rename tearDown to something like `executeTearDownForControl`

haozhun commented on presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java:
```diff
@@ -208,32 +214,74 @@ private boolean validate()
         // query has too many rows. Consider blacklisting.
         if (controlResult.getState() == State.TOO_MANY_ROWS) {
             testResult = new QueryResult(State.INVALID, null, null, null, null, ImmutableList.of());
+            tearDownControl();
```
```
boolean tearDownControl = true;
boolean tearDownTest = false;

try {
  ...
  tearDownTest = true;
  ...
  if (mismatch && noTearDownForMismatch) {
    tearDownControl/Test = false;
  }
}
finally {
  // tearDown
}

electrum commented on presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java:
```diff
@@ -297,71 +297,66 @@ private boolean checkForDeterministicAndRerunTestQueriesIfNeeded()
-            if (setupResult.getState() == State.SUCCESS) {
-                queryResult = executeQuery(controlGateway, controlUsername, controlPassword, queryPair.getControl(), query.getQuery(), controlTimeout, sessionProperties);
+            // startup
+            queryResult = setup(
```
This formatting is a bit hard to read. How about
```java
queryResult = setup(query, preQueryResults, preQuery ->
        executeQuery(gateway, username, password, query, preQuery, timeout, sessionProperties););

electrum commented on presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java:
```diff
@@ -351,39 +402,63 @@ private QueryResult executeQueries(
+        int attempt = 0;
+        QueryResult tearDownResult;
+        do {
+            tearDownResult = tearDown(
```
See earlier comment about formatting

electrum commented on presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java:
```diff
@@ -351,39 +402,63 @@ private QueryResult executeQueries(
+            }
+            try {
+                TimeUnit.MINUTES.sleep(1);
+                log.info(format("Query teardown failed on attempt #%d, will sleep and retry", attempt));
```
Always use `%s` in preference to a plain `%d` as it is safer

electrum replied:
`format()` is not needed for the logger as the logger supports formatting natively

electrum commented on presto-verifier/src/main/java/com/facebook/presto/verifier/Validator.java:
```diff
@@ -351,39 +402,63 @@ private QueryResult executeQueries(
+            attempt++;
+        }
+        while (attempt < teardownRetries);
+        return tearDown(
```
Same comment about formatting
